diff --git a/lib/Conversion/TTKernelToEmitC/TTKernelToEmitC.cpp b/lib/Conversion/TTKernelToEmitC/TTKernelToEmitC.cpp
index 8d7fd4a01..72fd79f7c 100644
--- a/lib/Conversion/TTKernelToEmitC/TTKernelToEmitC.cpp
+++ b/lib/Conversion/TTKernelToEmitC/TTKernelToEmitC.cpp
@@ -149,7 +149,7 @@ static void assignRuntimeCBArgIndices(func::FuncOp funcOp) {
 
 static bool isCBDeclaration(emitc::VerbatimOp verbatim,
                             llvm::StringRef cbName) {
-  std::string prefix = "CircularBuffer " + cbName.str() + "(";
+  std::string prefix = "experimental::CircularBuffer " + cbName.str() + "(";
   return verbatim.getValue().starts_with(prefix);
 }
 
@@ -206,7 +206,7 @@ static std::string ensureCBDeclaration(Value cb, Operation *useOp,
     rewriter.setInsertionPointToStart(cb.getParentBlock());
   }
 
-  std::string cbDecl = "CircularBuffer " + cbName + "({});";
+  std::string cbDecl = "experimental::CircularBuffer " + cbName + "({});";
   rewriter.create<emitc::VerbatimOp>(useOp->getLoc(), cbDecl, ValueRange{cb});
   return cbName;
 }
diff --git a/lib/Target/TTKernel/TTKernelToCpp.cpp b/lib/Target/TTKernel/TTKernelToCpp.cpp
index 485af5c4d..a24ca819b 100644
--- a/lib/Target/TTKernel/TTKernelToCpp.cpp
+++ b/lib/Target/TTKernel/TTKernelToCpp.cpp
@@ -176,8 +176,8 @@ public:
     region->walk([&](emitc::VerbatimOp verbatimOp) {
       llvm::StringRef value = verbatimOp.getValue();
 
-      if (value.starts_with("CircularBuffer")) {
-        headers.insert("api/dataflow/circular_buffer.h");
+      if (value.starts_with("experimental::CircularBuffer")) {
+        headers.insert("experimental/circular_buffer.h");
       }
 
       // Some callees are embedded in VerbatimOps.
